core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -202,6 +201,7 @@ public long spill(long size, MemoryConsumer trigger) throws IOException {
     // pages will currently be counted as memory spilled even though that space isn't actually
     // written to disk. This also counts the space needed to store the sorter's pointer array.
     taskContext.taskMetrics().incMemoryBytesSpilled(spillSize);
+    taskContext.taskMetrics().incDiskBytesSpilled(writeMetrics.shuffleBytesWritten());
```

Comment:
The `spillSize` here is 0 because the data are stored in a map instead of this sorter. So `incMemoryBytesSpilled(spillSize)` actually increase 0. We need update the `MemoryBytesSpilled` after freeing the memory in the map.